### ISSUE TITLE
Unable to know batch number from within custom train_step

### ISSUE THREAD

**Issue by alexbaras** [keras-team/keras#14483]
I am trying to implement a custom keras Model, specifically my own train_step method in which I am accumulating the gradients over ever n batches and then applying. I am having a very hard time trying to actually access the current batch number. I have tried setting attributes of the class, but they don't seem to be getting updated during training. I also tried via custom callback but no luck. I can get this all work in a custom training loop but would prefer to see if this could work with a custom keras Model.

Here is my current code. Any suggestions would be appreciated.

```python
import numpy as np
import tensorflow as tf
from matplotlib import pyplot as plt
from tqdm import tqdm
import tensorflow_datasets as tfds


class GradAccumModel(tf.keras.Model):
    def __init__(self, n_gradients, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.n_gradients = n_gradients
        self.accumulated_gradients = None
        self.apply_gradients = True

    # @tf.function
    def train_step(self, data):
        # Unpack the data. Its structure depends on your model and on what you pass to `fit()`
        x, y = data

        # Gradient tape
        with tf.GradientTape() as tape:
            # forward pass
            y_pred = self(x, training=True)  # Forward pass
            # Compute the loss value
            loss = self.compiled_loss(y, y_pred, regularization_losses=self.losses)

        # Compute batch gradients
        gradients = tape.gradient(loss, self.trainable_variables)
        # Apply batch gradients
        # self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))

        # Accumulate gradients
        if self.accumulated_gradients is None:
            self.accumulated_gradients = gradients
        else:
            for i in range(len(self.accumulated_gradients)):
                self.accumulated_gradients[i] = self.accumulated_gradients[i] + gradients[i]

        # Conditionally apply gradients 
        # TODO: find current batch number?
        if bool(((self.optimizer.iterations + 1) % self.n_gradients) == 0):
            for i in range(len(self.accumulated_gradients)):
                self.accumulated_gradients[i] = self.accumulated_gradients[i] / self.n_gradients
            self.optimizer.apply_gradients(zip(self.accumulated_gradients, self.trainable_variables))
            self.accumulated_gradients = None

        # Update metrics (includes the metric that tracks the loss)
        self.compiled_metrics.update_state(y, y_pred)
        # Return a dict mapping metric names to current value
        return {m.name: m.result() for m in self.metrics}


class ApplyGradient(tf.keras.callbacks.Callback):
    def on_train_batch_begin(self, batch, logs=None):
        if (batch + 1) % self.model.n_gradients == 0:
            self.model.apply_gradients = True
        else:
            self.model.apply_gradients = False


def build_model(input_spec, image_encoder, n_output, n_gradients):
    input_images = tf.keras.layers.Input(shape=input_spec._shape[1:], dtype=input_spec._dtype)
    images_encoded = image_encoder(input_images / 225)
    images_flatten = tf.keras.layers.Flatten()(images_encoded)
    hidden = tf.keras.layers.Dropout(0.1)(images_flatten)
    hidden = tf.keras.layers.Dense(256, activation=tf.keras.activations.relu)(hidden)
    image_probabilities = tf.keras.layers.Dense(n_output, activation=tf.keras.activations.softmax)(hidden)

    if n_gradients is not None:
        return GradAccumModel(n_gradients=n_gradients, inputs=[input_images], outputs=[image_probabilities])
    else:
        return tf.keras.Model(inputs=[input_images], outputs=[image_probabilities])


# get training and valid data
ds = tfds.load('cifar10', split=['train', 'test'], shuffle_files=True, with_info=True)
ds_train = ds[0][0].map(lambda elem: (elem['image'], tf.one_hot(elem['label'], depth=10, on_value=1, off_value=0, dtype=tf.float32)))
ds_train = ds_train.shuffle(10000, reshuffle_each_iteration=True).batch(64, drop_remainder=True)

# make model
tile_encoder = tf.keras.applications.ResNet50V2(input_shape=ds_train.element_spec[0]._shape[1:], include_top=False, weights='imagenet', pooling=None)
tile_encoder.trainable = False
model = build_model(ds_train.element_spec[0], tile_encoder, n_output=10, n_gradients=5)

# compile model
metrics = [tf.keras.metrics.CategoricalAccuracy(), tf.keras.metrics.AUC(num_thresholds=100)]
loss = [tf.keras.losses.CategoricalCrossentropy(from_logits=False)]
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001), loss=loss, metrics=metrics)

# callbacks = [ApplyGradient()]
model.fit(ds_train, epochs=10, verbose=1)
# for i in range(10):
#     for data in ds_train:
#         model.train_step(data)

hist = model.history.history
# plot training losses
_, ax = plt.subplots(ncols=2)
[ax[0].plot(hist[key], label=key) for key in hist]
```

**Comment by alexbaras** [comment]
I suspect this is because the train_function is cached

By further extending the .fit method as well I was able to achieve what was needed

The key point being that a tf.Variable with the boolean flag was needed to be set in .fit method.

I could not figure any other way to achieve this.

```python
import tensorflow as tf
import copy
from tensorflow.python.eager import backprop, context, def_function
from tensorflow.python.keras.engine import data_adapter, training_utils
from tensorflow.python.keras.engine.training import enable_multi_worker, reduce_per_replica, _keras_api_gauge
from tensorflow.python.keras.utils import version_utils
from tensorflow.python.keras import callbacks as callbacks_module
from tensorflow.python.profiler import traceme


class GradientAccumulationModel(tf.keras.Model):
    def __init__(self, n_gradients, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.n_gradients = n_gradients
        self.gradient_accumulation = [tf.Variable(tf.zeros_like(v, dtype=tf.float32), trainable=False) for v in self.trainable_variables]
        self.apply_gradients = tf.Variable(False, dtype=tf.bool, trainable=False)
        self.train_function = None

    def train_step(self, data, apply_gradients=False):
        # Unpack data
        data = data_adapter.expand_1d(data)
        x, y, sample_weight = data_adapter.unpack_x_y_sample_weight(data)

        # Gradient Tape
        with backprop.GradientTape() as tape:
            # Forward pass
            y_pred = self(x, training=True)
            # Loss
            loss = self.compiled_loss(y, y_pred, sample_weight, regularization_losses=self.losses)

        # Calculate batch gradients
        batch_gradients = tape.gradient(loss, self.trainable_variables)
        # Accumulate batch gradients
        for i in range(len(self.gradient_accumulation)):
            self.gradient_accumulation[i].assign_add(batch_gradients[i])

        # Apply accumulated gradients, conditionally
        if apply_gradients:
            # apply
            self.optimizer.apply_gradients(zip(self.gradient_accumulation, self.trainable_variables))
            # reset
            for i in range(len(self.gradient_accumulation)):
                self.gradient_accumulation[i].assign(tf.zeros_like(self.trainable_variables[i], dtype=tf.float32))

        # Update metrics
        self.compiled_metrics.update_state(y, y_pred, sample_weight)
        return {m.name: m.result() for m in self.metrics}

    def make_train_function(self):
        if self.train_function is not None:
            return self.train_function

        def train_function(iterator):
            data = next(iterator)
            outputs = self.distribute_strategy.run(self.train_step, args=(data, self.apply_gradients))
            outputs = reduce_per_replica(outputs, self.distribute_strategy, reduction='first')
            return outputs

        if not self.run_eagerly:
            train_function = def_function.function(train_function, experimental_relax_shapes=True)

        self.train_function = train_function
        return self.train_function

    @enable_multi_worker
    def fit(self,
            x=None,
            y=None,
            batch_size=None,
            epochs=1,
            verbose=1,
            callbacks=None,
            validation_split=0.,
            validation_data=None,
            shuffle=True,
            class_weight=None,
            sample_weight=None,
            initial_epoch=0,
            steps_per_epoch=None,
            validation_steps=None,
            validation_batch_size=None,
            validation_freq=1,
            max_queue_size=10,
            workers=1,
            use_multiprocessing=False):

        _keras_api_gauge.get_cell('fit').set(True)
        # Legacy graph support is contained in `training_v1.Model`.
        version_utils.disallow_legacy_graph('Model', 'fit')
        self._assert_compile_was_called()
        self._check_call_args('fit')

        if validation_split:
            # Create the validation data using the training data. Only supported for
            # `Tensor` and `NumPy` input.
            (x, y, sample_weight), validation_data = (
                data_adapter.train_validation_split((x, y, sample_weight),
                                                    validation_split=validation_split,
                                                    shuffle=False))

        with self.distribute_strategy.scope(), training_utils.RespectCompiledTrainableState(self):
            # Creates a `tf.data.Dataset` and handles batch and epoch iteration.
            data_handler = data_adapter.DataHandler(
                x=x,
                y=y,
                sample_weight=sample_weight,
                batch_size=batch_size,
                steps_per_epoch=steps_per_epoch,
                initial_epoch=initial_epoch,
                epochs=epochs,
                shuffle=shuffle,
                class_weight=class_weight,
                max_queue_size=max_queue_size,
                workers=workers,
                use_multiprocessing=use_multiprocessing,
                model=self)

            # Container that configures and calls `tf.keras.Callback`s.
            if not isinstance(callbacks, callbacks_module.CallbackList):
                callbacks = callbacks_module.CallbackList(
                    callbacks,
                    add_history=True,
                    add_progbar=verbose != 0,
                    model=self,
                    verbose=verbose,
                    epochs=epochs,
                    steps=data_handler.inferred_steps)

            self.stop_training = False
            train_function = self.make_train_function()
            callbacks.on_train_begin()
            # Handle fault-tolerance for multi-worker.
            # TODO(omalleyt): Fix the ordering issues that mean this has to
            # happen after `callbacks.on_train_begin`.
            data_handler._initial_epoch = (  # pylint: disable=protected-access
                self._maybe_load_initial_epoch_from_ckpt(initial_epoch))
            for epoch, iterator in data_handler.enumerate_epochs():
                self.reset_metrics()
                callbacks.on_epoch_begin(epoch)
                with data_handler.catch_stop_iteration():
                    for step in data_handler.steps():
                        with traceme.TraceMe(
                                'TraceContext',
                                graph_type='train',
                                epoch_num=epoch,
                                step_num=step,
                                batch_size=batch_size):
                            callbacks.on_train_batch_begin(step)
                            self.apply_gradients.assign(((step + 1) % self.n_gradients) == 0)
                            tmp_logs = train_function(iterator)
                            # Catch OutOfRangeError for Datasets of unknown size.
                            # This blocks until the batch has finished executing.
                            # TODO(b/150292341): Allow multiple async steps here.
                            if not data_handler.inferred_steps:
                                context.async_wait()
                            logs = tmp_logs  # No error, now safe to assign to logs.
                            callbacks.on_train_batch_end(step, logs)
                epoch_logs = copy.copy(logs)

                # Run validation.
                if validation_data and self._should_eval(epoch, validation_freq):
                    val_x, val_y, val_sample_weight = (
                        data_adapter.unpack_x_y_sample_weight(validation_data))
                    val_logs = self.evaluate(
                        x=val_x,
                        y=val_y,
                        sample_weight=val_sample_weight,
                        batch_size=validation_batch_size or batch_size,
                        steps=validation_steps,
                        callbacks=callbacks,
                        max_queue_size=max_queue_size,
                        workers=workers,
                        use_multiprocessing=use_multiprocessing,
                        return_dict=True)
                    val_logs = {'val_' + name: val for name, val in val_logs.items()}
                    epoch_logs.update(val_logs)

                callbacks.on_epoch_end(epoch, epoch_logs)
                if self.stop_training:
                    break

            callbacks.on_train_end()
            return self.history

```

**Comment by alexbaras** [comment]
I will close this as technically it is "resolved" at this point

**Comment by JanKaifer** [comment]
For other googlers:

I used `tf.Variable` initialized in model constructor and incremented in `train_step` with:
```python
self.step_var.assign_add(delta=1)
```
